### PR TITLE
fixes to the toc formatting

### LINKED ISSUE #9783: Create Book (Preview) doesn't work for notebooks spanning across multiple folders (list index out of range)
- Azure Data Studio Version:
> Version: 1.17.0-insider (user setup)
> Commit: be83b31e378774ed14879fb1889c47422e68bea5
> Date: 2020-03-27T06:21:06.051Z
> VS Code: 1.42.0
> Electron: 7.1.11
> Chrome: 78.0.3904.130
> Node.js: 12.8.1
> V8: 7.8.279.23-electron.0
> OS: Windows_NT x64 10.0.18363

Steps to Reproduce:

1. Setup the directory of the notebook files to be imported (source files) in multiple directories.
NotebookSource\Chapter1\Notebook11.ipynb
NotebookSource\Chapter1\Notebook12.ipynb
NotebookSource\Chapter1\Notebook13.ipynb
NotebookSource\Chapter2\Notebook21.ipynb
NotebookSource\Chapter2\Notebook22.ipynb
NotebookSource\Chapter2\Notebook23.ipynb

2. Use Create Book (Preview) and follow the instructions.
On the third cell you'll see "list index out of range"

image

Expected outcome:
- MyNewNotebook
   - Chapter 1
        - Notebook11.ipynb
        - Notebook12.ipynb
        - Notebook13.ipynb
   - Chapter 2
        - Notebook21.ipynb
        - Notebook22.ipynb
        - Notebook23.ipynb

### PR DESCRIPTION
**Pull request by MaddyDev**
Before Fix (Issue described in #9780): Create Book assumes that there should be a markdown file in each folder and creates a tree structure and in the absence of the markdown, one of the ipynb files in the given list is taken as preview file and a tree structure is created. 
Fix: If and only when there is a markdown, it creates a preview file and tree structure, otherwise it just adds the files under the book.
Second issue #9783: Create Book doesn't work for notebooks spanning across multiple folders. 
This is also dependent on the first issue and in the absence of markdown file, due to the prior assumptions I made, the code errors out. 
Fix: Added proper checks for existence of markdown file.

This PR fixes #9783 and #9780 

In the below screenshot: Book3 has markdown "areadme" file, with preview and notebooks under it. Book4 doesn't have a markdown, so the tree structure isn't there anymore.
image

### CODE REVIEW

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 1]
```diff
@@ -97,10 +97,25 @@
...
             "execution_count": null
         },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## 3. Format toc.yml file of the book\r\n",
+                "\r\n",
+                "Create a valid toc.yml file that enables your book to be opened on our view-let.\r\n",
+                "\r\n",
+                "<span style=\"color:red\">Note:</span> Skipping this step would have ill-formed toc and need your manual updates to it."
```
Comment (corivera) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:107]: Might want to reword this for greater clarity. People using this notebook probably wouldn't know anything about the Book file structure.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 2]
```diff
@@ -129,15 +152,36 @@
                 "        if (headers):\n",
                 "            for header in headers:\n",
                 "                title = header[10:].rstrip()\n",
-                "                contentFolders.append(title.lower())\n",
-                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x, urls))\n",
+                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x.lower(), urls))\n",
                 "                index = urls.index(filtered[len(filtered)-1])\n",
+                "                markdownFilePath = getMarkdownFile(filtered[len(filtered)-1])\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:157]: In python you should be able to just do `filtered[-1]` to access the last item in a list

Comment (Charles-Gagnon) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:157]: (there's a whole bunch of places where this is being done - that might be worth doing in a separate PR though to keep this one a bit cleaner)

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 3]
```diff
@@ -129,15 +152,36 @@
...
                 "                result = result.replace(header, replacedString)\n",
-                "                result = result.replace(urls[index], '')\n",
-                "                del urls[index]\n",
-                "        if (urls):\n",
+                "                if (markdownExists):\n",
+                "                    result = result.replace(urls[index], '')\n",
+                "                    del urls[index]\n",
+                "            if (urls):\n",
+                "                for url in urls:\n",
+                "                    title = url[url.rindex(os.path.sep)+1:].rstrip()\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:172]: You're just trying to get the filename out of this right? Does path.split/path.basename not work for that? Doing manual string parsing like this isn't great...

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 4]
```diff
@@ -129,15 +152,36 @@
                 "        if (headers):\n",
                 "            for header in headers:\n",
                 "                title = header[10:].rstrip()\n",
-                "                contentFolders.append(title.lower())\n",
-                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x, urls))\n",
+                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x.lower(), urls))\n",
                 "                index = urls.index(filtered[len(filtered)-1])\n",
+                "                markdownFilePath = getMarkdownFile(filtered[len(filtered)-1])\n",
+                "                markdownExists = path.exists(markdownFilePath)\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:158]: In general please add more comments explaining what each step is doing. It's very unclear what all this logic is for which makes it a lot harder to determine if it's doing the right thing.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 5]
```diff
@@ -97,10 +97,25 @@
...
+            "metadata": {
+                "azdata_cell_guid": "5a29e2dd-d9f5-4dc1-87bc-bce639362f51"
+            }
+        },
         {
             "cell_type": "code",
             "source": [
                 "# Update toc file, book title and clean up the directores\n",
+                "\n",
+                "from os import path\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:118]: You have a mix of os.path and just path. I'd suggest just importing path like this up above and then converting all the os.path in the notebook to just use path since that's a lot cleaner (but do that in a separate PR - maybe jus tmake one with the negative indexing/other cleanup too)

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 6]
```diff
@@ -129,15 +152,36 @@
                 "        if (headers):\n",
                 "            for header in headers:\n",
                 "                title = header[10:].rstrip()\n",
-                "                contentFolders.append(title.lower())\n",
-                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x, urls))\n",
+                "                filtered = list(filter(lambda x: (\"%s%s%s\" % (os.path.sep, title.lower(), os.path.sep)) in x.lower(), urls))\n",
```
[the remark this answers is not in the document]
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb]: not an expert in python either, just googled online and added this snippet for filtering out an expression in an array.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 2 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:187]: didn't check if that works, will do it in a seperate PR. Thanks.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 3 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb]: the prob is url is not path it's string which has data along with path. For ex: (- url: DsMain\content\Scenario1\1.MakeYourFirstCodeChange) Thus all the fuss.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 4 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:188]: I did add some comments around, but I still need to do better. Most of my time is gone fiddling around python code since am new it. But I'll make sure add some more comments in a separate PR.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 7]
```diff
@@ -71,24 +71,27 @@
             "source": [
                 "import os, re, shutil\r\n",
                 "\r\n",
-                "overwrite = False\r\n",
-                "book_name = input('Please provide the path where the book needs to be saved along with the book name ex-> D:\\Book1: ') \r\n",
+                "try:\r\n",
+                "    overwrite = False\r\n",
+                "    book_name = input('Please provide the path where the book needs to be saved along with the book name ex-> D:\\Book1: ') \r\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:76]: This would be a lot better if we could use a file picker dialog - have we done anything like that before from a Notebook?

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 8]
```diff
@@ -112,54 +130,108 @@
                 "firstLevelUrls = re.findall(r'^(?:\\s+$[\\r\\n]+)+(\\- url: [a-zA-Z0-9\\\\.\\s\\-\\/]+$[\\r\\n]+)', result, re.MULTILINE)\n",
                 "urls = re.findall(r'- url: [a-zA-Z0-9\\\\.\\s\\-\\/]+$', result, re.MULTILINE)\n",
                 "headers = re.findall(r'- header: [a-zA-Z0-9\\\\.\\s-]+$', result, re.MULTILINE)\n",
+                "# all the markdown urls are placed at the end of the list. \n",
+                "possibleMarkdowns = re.findall(r'(- url: [a-zA-Z0-9\\\\.\\s\\-\\/]+$[\\r\\n]+)[\\r\\n]', result, re.MULTILINE)\n",
+                "\n",
+                "def getMarkdownFile(url):\n",
+                "    # url are usually defined in toc as: \"- url: <Path\\to\\notebookFile>\"\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:137]: Are "usually"? That implies that sometimes they wouldn't be which would break this...is that actually the case? Also I didn't check all the code but you making sure to trim the string - or is this guaranteed not to start with spaces?

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb [hunk 9]
```diff
@@ -112,54 +130,108 @@
...
+                "    # substring from 7th postion excluding the \"- url: \" from the path\n",
+                "    #folders = url[7:].rstrip().split(path.sep)\n",
+                "    if (os.name == 'nt'):\n",
+                "        # if windows get the drive letter and add it to the url\n",
+                "        driveletter = content_folder.split(path.sep)[0]\n",
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:142]: Do we support UNC paths for this?

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 7 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:76]: Will look into that. Thanks.

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 8 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:137]: So currently that's how Jupyter formats the toc.yml unless they change it in the future. Probably mush have written "Currently" instead of "usually"

File: extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb (hunk 9 above)
Comment (MaddyDev) [reply extensions/notebook/resources/notebooks/JupyterBooksCreate.ipynb:142]: no we don't, but this is one more issue that I've seen with jupyter formatting the toc file. It removed the drive letter and specifies the urls.

**Comment by MaddyDev** [comment]
I'll look into file picker dialog in a separate PR. Will get this in and let it be tested.

> I didn't spend a lot of time going over the logic in fine detail - it's such a mess to try and read this stuff that I think we're better off just having people test it and fix any issues that come up then.

I agree, am not an expert on how I can modularize this python code for easier readability without splitting them into multiple code cells and lengthening the notebook. I've tested this and gave it to Julie to test out her scenarios, so far we haven't run into issues. So I'll get this in for further testing in through Insiders. Since we're moving to dialog based creation in our near future, I think spending more time here is not a good idea.
